### ISSUE TITLE
Allow to override the default description in the help-text

### ISSUE THREAD

**Issue by ModernRonin** [ModernRonin/FluentArgumentParser#28]
for situations where the type of the property is nullable (string), maybe even enforce this

**Comment by victormarante** [comment]
Hi, I wanted to contribute to this solution, ended up cloning the project. However, I get a lot of build errors. Is there something I need to know before being able to start coding?

**Comment by victormarante** [comment]
> Hi there :-)
> 
> 
> 
> The project uses [paket](https://fsprojects.github.io/Paket/) to manage nugets.
> 
> 
> 
> so you need to do (in the directory of the solution, from the command-line):
> 
> ```
> 
> dotnet tool restore
> 
> dotnet paket restore
> 
> ```
> 
> that will install paket and restore the packages. From this point on, you can work with it quite normally, unless you need to add a nuget package. (You probably don't need that, but if you do, you'll need to read the paket docs a bit.)
> 
> 
> 
> 
> 
> if you run into anything else, let me know.
> 
> 
> 
> 
> 
> 

Hi, thanks for the quick response. Great, I will download and use Paket and see if my errors go away.

Maybe add a part about this in the Contribution part of README.md file? I can make a PR updating the documentation if you'd like.

**Comment by ModernRonin** [comment]
Yeah, you are right, it's just that so far you're the first who showed any interest in contributing, so I hadn't noticed the need yet.

If you can do the PR for that, that would be awesome.

Out of curiosity: how did you get the idea to contribute? are you a user of the lib and want to do this issue because it applies to your use-case?

**Comment by victormarante** [comment]
I will message you privately to not bloat issue.

**Comment by victormarante** [comment]
I will familiarise myself with the code, and start working on this issue. Thanks for the help.

**Comment by ModernRonin** [comment]
thanks for wanting to contribute!

**Comment by victormarante** [comment]
Feel free to assign this issue to me whenever you have the time.

**Comment by ModernRonin** [comment]
done. FYI, typically I'll be able to react between 0900-1000 and 1800-2100 UTC.

**Comment by victormarante** [comment]
> 

Good to know!

**Comment by victormarante** [comment]
> done. FYI, typically I'll be able to react between 0900-1000 and 1800-2100 UTC.

Good to know!

**Comment by victormarante** [comment]
@ModernRonin Hi! It's probably going to take some time for me to understand your code. It is quite complex (for me), and I am not entirely sure what this issue entitles, just as an FYI.

**Comment by ModernRonin** [comment]
You'll need to look at:
* `HelpMaker.Format()`, specifically the local function `optionalToRow`; currently the help text is simply the default value of the parameter
* `ParameterBindingConfigurer`; this contains all the fluent methods for describing a parameter; here you need to add a method `.WithDefaultDescription(string text)`; now at this level you will have to check whether `AParameter` is actually optional. see the method `.WithDefault()` for an example

you probably want to add a property to `OptionalParameter` to hold the customized description; in the new method of `ParameterBindingConfigurer` you set it, in `HelpMaker` you use it if it's set, otherwise stay with the behavior as is


hope that helps :-)

**Comment by ModernRonin** [comment]
(these infos are enough to allow what the title of the issue says; for the second part, described in the first comment "maybe enforce this when the parameter type is a reference type with a null default", you'd look at `OptionalParameterValidator`. (validation uses [FluentValidation](https://fluentvalidation.net/), so if you are not familiar with it, check out the link for docs.)

**Comment by victormarante** [comment]
> You'll need to look at:
> 
> * `HelpMaker.Format()`, specifically the local function `optionalToRow`; currently the help text is simply the default value of the parameter
> * `ParameterBindingConfigurer`; this contains all the fluent methods for describing a parameter; here you need to add a method `.WithDefaultDescription(string text)`; now at this level you will have to check whether `AParameter` is actually optional. see the method `.WithDefault()` for an example
> 
> you probably want to add a property to `OptionalParameter` to hold the customized description; in the new method of `ParameterBindingConfigurer` you set it, in `HelpMaker` you use it if it's set, otherwise stay with the behavior as is
> 
> hope that helps :-)

Hi, thanks a ton. This will definetely be enough to get me started.

> (these infos are enough to allow what the title of the issue says; for the second part, described in the first comment "maybe enforce this when the parameter type is a reference type with a null default", you'd look at `OptionalParameterValidator`. (validation uses [FluentValidation](https://fluentvalidation.net/), so if you are not familiar with it, check out the link for docs.)

Thanks! I will look into this today.

**Comment by victormarante** [comment]
#32

**Comment by ModernRonin** [comment]
please check the CI run in the PR

**Comment by ModernRonin** [comment]
also, my other comment there

**Comment by victormarante** [comment]
Answered in PR.
